packages/sdk/contractkit/src/web3-contract-cache.ts

```diff
@@ -155,17 +158,24 @@ export class Web3ContractCache {
    */
   async getContract<C extends keyof typeof ContractFactories>(contract: C, address?: string) {
     if (this.cacheMap[contract] == null || address !== undefined) {
+      // core contract in the registry
+      if (!address) {
```

Comment:
this isn't strictly about this change but trying to understand how this function is used a bit better -- is there a reason we don't enforce the XOR? I'm imagining a case where a user passes in `contract` and the wrong `address` for that contract and unintentionally caches the wrong mapping leading to confusing behavior. Is that actually possible here or am I missing something?

Reply:
This function has the possibility of sending an address just because we have a few contracts that are not from the registry, but could be deployed in multiple addresses (like MetaTransactionWallet or MultiSig). Probably (just guessing), this method started as a private one, and then changed to be public. In a way we could check if only the "Not in registry"-core-contracts have an address, but probably is easier to debug a new version of a core contract that is deployed somewhere else (that's why I think maybe this is a public one).

The `await this.kit.registry.addressFor(contract)` it actually throws an error if the address in the registry is the NULL_ADDRESS. What I tried here, is to change that message for something that could lead the user to understand that is not its fault, but just the state of the chain.

And as you said, it is possible for a user to send a invalid address with a core contract, I don't really know right now if contract creation in the line 176 sends something to check that the abi matches with the one from the address. If that's the case, it will fail. If it's not (pretty sure this is the case), as you pointed, the user will catch a wrong contract, and all the calls to that contract will fail.
In a way, if the user is actually changing the address, instead of using the getters, probably will know it could be its fault if everything is broken. We could have some check there, but, IMHO it will add a call in every contract creation, just in case the user changed the address (which should be really uncommon)

Reply:
I see what you mean, I think a simpler check/heuristic (of course not as robust as checking both ABIs as you mentioned) would be to just ensure that the user only passes in either the contract name or an address but not both (i.e. it would fail or, if we want to add some help without losing backwards compatibility, warn if they pass in both args)

Reply:
The thing is that you require the Contract name to see which wrapper should be used, for the ones that does not live in the registry, MultiSig, for example.
The multisig will require both

Reply:
ahh ok gotcha, didn't understand that use case before. Thanks for the clarification, looks good to me as is then! :)